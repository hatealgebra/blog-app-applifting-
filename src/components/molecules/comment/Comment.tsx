import React from 'react';
import { RiArrowUpSLine } from '@react-icons/all-files/ri/RiArrowUpSLine';
import { RiArrowDownSLine } from '@react-icons/all-files/ri/RiArrowDownSLine';
import { Components } from '@customTypes/declarations';
import { timeDifference } from '../../../utils/date.utils';
import Avatar from '../../atoms/avatar/Avatar';
import { StyledCommentContainer, StyledCommentCounter } from './comment.styled';

import { voteDown, voteUp } from '../../../services/commentsServices';

const Comment = ({
  commentId,
  author,
  score,
  postedAt,
  content,
}: Components['schemas']['Comment']) => {
  const [counterValue, setCounterValue] = React.useState(score ?? 0);
  const timestampNow = Date.now();

  const increaseScore = async () => {
    try {
      voteUp(commentId!);
      return setCounterValue((prev) => prev + 1);
    } catch (e) {
      return e;
    }
  };

  const decreaseScore = async () => {
    try {
      voteDown(commentId!);
      return setCounterValue((prev) => prev - 1);
    } catch (e) {
      return e;
    }
  };

  return (
    <StyledCommentContainer className="comment">
      <Avatar size="lg" />
      <h5 className="comment__author">{author}</h5>
      <span className="comment__time-diff label">
        {timeDifference(timestampNow, postedAt!)}
      </span>
      <p className="comment__text">{content}</p>
      <StyledCommentCounter>
        <span data-testid="reactionCounter">
          {counterValue! > 0 ? '+' : ''}
          {counterValue}
        </span>
        <button aria-label="vote-up" onClick={increaseScore}>
          <RiArrowUpSLine size="100%" />
        </button>
        <button aria-label="vote-down" onClick={decreaseScore}>
          <RiArrowDownSLine size="100%" />
        </button>
      </StyledCommentCounter>
    </StyledCommentContainer>
  );
};

export default Comment;
