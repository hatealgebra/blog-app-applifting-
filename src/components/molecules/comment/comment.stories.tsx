import React from 'react';
import { Components } from '@customTypes/declarations';
import { userEvent, within } from '@storybook/testing-library';

import Comment from './Comment';

const CommentTemplate = (args: Components['schemas']['Comment']) => (
  <Comment
    {...args}
    articleId="hello world"
    author="Linda Lynn"
    score={0}
    postedAt="2020-10-01T19:44:11.577668"
    content="Lorem ipsum dolor, sit amet consectetur adipisicing elit. Natus quibusdam qui autem libero consequuntur tenetur eveniet recusandae aliquam obcaecati.Consequatur perferendis id, dolorem quae unde recusandae ad delenitiobcaecati similique?"
  />
);

export const CommentExample = CommentTemplate.bind({});
CommentExample.args = {
  articleId: 'hello world',
  author: 'Linda Lynn',
  score: 0,
  postedAt: '2020-10-01T19:44:11.577668',
  content:
    'Lorem ipsum dolor, sit amet consectetur adipisicing elit. Natus quibusdam qui autem libero consequuntur tenetur eveniet recusandae aliquam obcaecati.Consequatur perferendis id, dolorem quae unde recusandae ad delenitiobcaecati similique?',
};

export const CommentVotedUp = CommentTemplate.bind({});
CommentVotedUp.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const voteUpButton = canvas.getByRole('button', { name: 'vote-up' });
  await userEvent.click(voteUpButton);
};

export const CommentVoteDown = CommentTemplate.bind({});
CommentVoteDown.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const voteDownButton = canvas.getByRole('button', { name: 'vote-down' });
  await userEvent.click(voteDownButton);
};

// export const CommentAPIError = CommentTemplate.bind({});
// CommentAPIError.parameters = {
//   msw: [
//     rest.post(
//       `${BASE_API_URL}/comments/:commentId/vote/up`,
//       (req, res, ctx) => {
//         return res(ctx.status(500));
//       }
//     ),
//   ],
// };

export default {
  title: 'Molecules/Comment',
  component: Comment,
};
