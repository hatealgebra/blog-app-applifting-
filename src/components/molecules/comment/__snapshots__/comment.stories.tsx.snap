// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/Comment CommentExample smoke-test 1`] = `
<div class="sc-bjEwCx jyGvnB comment">
  <img class="sc-bcPKhP jnhZlJ avatar"
       src="static/media/src/images/avatar-fallback.png"
       alt="fallback user avatar"
  >
  <h5 class="comment__author">
    Linda Lynn
  </h5>
  <span class="comment__time-diff label">
    3 years ago
  </span>
  <p class="comment__text">
    Lorem ipsum dolor, sit amet consectetur adipisicing elit. Natus quibusdam qui autem libero consequuntur tenetur eveniet recusandae aliquam obcaecati. Consequatur perferendis id, dolorem quae unde recusandae ad deleniti obcaecati similique?
  </p>
  <div class="sc-fEyylQ gCWoPR comment__voting-counter">
    <span data-testid="reactionCounter">
      0
    </span>
    <button aria-label="vote-up">
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           height="100%"
           width="100%"
           xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path fill="none"
                d="M0 0h24v24H0z"
          >
          </path>
          <path d="M12 10.828l-4.95 4.95-1.414-1.414L12 8l6.364 6.364-1.414 1.414z">
          </path>
        </g>
      </svg>
    </button>
    <button aria-label="vote-down">
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 24 24"
           height="100%"
           width="100%"
           xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path fill="none"
                d="M0 0h24v24H0z"
          >
          </path>
          <path d="M12 13.172l4.95-4.95 1.414 1.414L12 16 5.636 9.636 7.05 8.222z">
          </path>
        </g>
      </svg>
    </button>
  </div>
</div>
`;
